File: templates/wasm-node-python/examples/control_flow.py
"""
Control Flow Nodes — Logic and branching operations

Demonstrates if/else branching, comparison, boolean gates,
conditional pass-through, and sequencing.
"""

from sdk import (
    Context,
    ExecutionResult,
    NodeDefinition,
    PinDefinition,
    PinType,
)


def get_definitions() -> list[NodeDefinition]:
    nodes: list[NodeDefinition] = []

    # If Branch
    if_branch = NodeDefinition("if_branch_py", "If Branch", "Branches based on boolean condition", "Control/Branch")
    if_branch.add_pin(PinDefinition.input_exec("exec"))
    if_branch.add_pin(PinDefinition.input_pin("condition", PinType.BOOL, default=False))
    if_branch.add_pin(PinDefinition.output_exec("true"))
    if_branch.add_pin(PinDefinition.output_exec("false"))
    nodes.append(if_branch)

    # Compare
    compare = NodeDefinition("compare_py", "Compare", "Compares two values", "Control/Logic")
    compare.add_pin(PinDefinition.input_exec("exec"))
    compare.add_pin(PinDefinition.input_pin("a", PinType.F64, default=0.0))
    compare.add_pin(PinDefinition.input_pin("b", PinType.F64, default=0.0))
    compare.add_pin(PinDefinition.output_exec("exec_out"))
    compare.add_pin(PinDefinition.output_pin("equal", PinType.BOOL))
    compare.add_pin(PinDefinition.output_pin("less_than", PinType.BOOL))
    compare.add_pin(PinDefinition.output_pin("greater_than", PinType.BOOL))
    nodes.append(compare)

    # AND Gate
    and_gate = NodeDefinition("and_gate_py", "AND Gate", "Logical AND of two booleans", "Control/Logic")
    and_gate.add_pin(PinDefinition.input_exec("exec"))
    and_gate.add_pin(PinDefinition.input_pin("a", PinType.BOOL, default=False))
    and_gate.add_pin(PinDefinition.input_pin("b", PinType.BOOL, default=False))
    and_gate.add_pin(PinDefinition.output_exec("exec_out"))
    and_gate.add_pin(PinDefinition.output_pin("result", PinType.BOOL))
    nodes.append(and_gate)

    # OR Gate
    or_gate = NodeDefinition("or_gate_py", "OR Gate", "Logical OR of two booleans", "Control/Logic")
    or_gate.add_pin(PinDefinition.input_exec("exec"))
    or_gate.add_pin(PinDefinition.input_pin("a", PinType.BOOL, default=False))
    or_gate.add_pin(PinDefinition.input_pin("b", PinType.BOOL, default=False))
    or_gate.add_pin(PinDefinition.output_exec("exec_out"))
    or_gate.add_pin(PinDefinition.output_pin("result", PinType.BOOL))
    nodes.append(or_gate)

    # NOT Gate
    not_gate = NodeDefinition("not_gate_py", "NOT Gate", "Logical NOT", "Control/Logic")
    not_gate.add_pin(PinDefinition.input_exec("exec"))
    not_gate.add_pin(PinDefinition.input_pin("value", PinType.BOOL, default=False))
    not_gate.add_pin(PinDefinition.output_exec("exec_out"))
    not_gate.add_pin(PinDefinition.output_pin("result", PinType.BOOL))
    nodes.append(not_gate)

    # Gate (conditional pass-through)
    gate = NodeDefinition("gate_py", "Gate", "Passes execution only if condition is true", "Control/Branch")
    gate.add_pin(PinDefinition.input_exec("exec"))
    gate.add_pin(PinDefinition.input_pin("condition", PinType.BOOL, default=True))
    gate.add_pin(PinDefinition.output_exec("exec_out"))
    nodes.append(gate)

    # Sequence
    sequence = NodeDefinition("sequence_py", "Sequence", "Activates multiple outputs in order", "Control/Flow")
    sequence.add_pin(PinDefinition.input_exec("exec"))
    sequence.add_pin(PinDefinition.output_exec("out_1"))
    sequence.add_pin(PinDefinition.output_exec("out_2"))
    sequence.add_pin(PinDefinition.output_exec("out_3"))
    nodes.append(sequence)

    return nodes


def run_if_branch(ctx: Context) -> ExecutionResult:
    condition = ctx.get_bool("condition", False)
    if condition:
        ctx.activate_exec("true")
    else:
        ctx.activate_exec("false")
    return ctx.finish()


def run_compare(ctx: Context) -> ExecutionResult:
    a = ctx.get_f64("a", 0.0)
    b = ctx.get_f64("b", 0.0)
    ctx.set_output("equal", a == b)
    ctx.set_output("less_than", a < b)
    ctx.set_output("greater_than", a > b)
    return ctx.success()


def run_and_gate(ctx: Context) -> ExecutionResult:
    a = ctx.get_bool("a", False)
    b = ctx.get_bool("b", False)
    ctx.set_output("result", a and b)
    return ctx.success()


def run_or_gate(ctx: Context) -> ExecutionResult:
    a = ctx.get_bool("a", False)
    b = ctx.get_bool("b", False)
    ctx.set_output("result", a or b)
    return ctx.success()


def run_not_gate(ctx: Context) -> ExecutionResult:
    value = ctx.get_bool("value", False)
    ctx.set_output("result", not value)
    return ctx.success()


def run_gate(ctx: Context) -> ExecutionResult:
    condition = ctx.get_bool("condition", True)
    if condition:
        return ctx.success()
    return ctx.finish()


def run_sequence(ctx: Context) -> ExecutionResult:
    ctx.activate_exec("out_1")
    ctx.activate_exec("out_2")
    ctx.activate_exec("out_3")
    return ctx.finish()


DISPATCH = {
    "if_branch_py": run_if_branch,
    "compare_py": run_compare,
    "and_gate_py": run_and_gate,
    "or_gate_py": run_or_gate,
    "not_gate_py": run_not_gate,
    "gate_py": run_gate,
    "sequence_py": run_sequence,
}


def run(node_name: str, ctx: Context) -> ExecutionResult:
    handler = DISPATCH.get(node_name)
    if handler is None:
        return ctx.fail(f"Unknown node: {node_name}")
    return handler(ctx)

File: templates/wasm-node-assemblyscript/examples/control_flow.ts
import {
	DataType,
	ExecutionResult,
	NodeDefinition,
	PinDefinition,
	Context,
	FlowNode,
	NodePackage,
} from "@flow-like/wasm-sdk/assembly/index";

export { alloc, dealloc, get_abi_version } from "@flow-like/wasm-sdk/assembly/index";

export class IfBranchNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "if_branch_as";
		def.friendly_name = "If Branch (AS)";
		def.description = "Routes execution based on a condition";
		def.category = "Control Flow";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("condition", "Condition", "Branch condition", DataType.Bool).withDefaultBool(false));
		def.addPin(PinDefinition.output("true_branch", "True", "Executes when true", DataType.Exec));
		def.addPin(PinDefinition.output("false_branch", "False", "Executes when false", DataType.Exec));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		if (ctx.getBool("condition")) {
			ctx.activateExec("true_branch");
		} else {
			ctx.activateExec("false_branch");
		}
		return ctx.finish();
	}
}

export class CompareNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "compare_as";
		def.friendly_name = "Compare Numbers (AS)";
		def.description = "Compares two numbers and outputs comparison results";
		def.category = "Control Flow";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First number", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.input("b", "B", "Second number", DataType.F64).withDefaultF64(0.0));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("equal", "Equal", "A equals B", DataType.Bool));
		def.addPin(PinDefinition.output("greater", "Greater", "A > B", DataType.Bool));
		def.addPin(PinDefinition.output("less", "Less", "A < B", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		const a = ctx.getF64("a");
		const b = ctx.getF64("b");
		ctx.setBool("equal", a == b);
		ctx.setBool("greater", a > b);
		ctx.setBool("less", a < b);
		return ctx.success();
	}
}

export class AndGateNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "and_gate_as";
		def.friendly_name = "AND Gate (AS)";
		def.description = "Logical AND of two boolean inputs";
		def.category = "Control Flow/Logic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First boolean", DataType.Bool).withDefaultBool(false));
		def.addPin(PinDefinition.input("b", "B", "Second boolean", DataType.Bool).withDefaultBool(false));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "A AND B", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setBool("result", ctx.getBool("a") && ctx.getBool("b"));
		return ctx.success();
	}
}

export class OrGateNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "or_gate_as";
		def.friendly_name = "OR Gate (AS)";
		def.description = "Logical OR of two boolean inputs";
		def.category = "Control Flow/Logic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First boolean", DataType.Bool).withDefaultBool(false));
		def.addPin(PinDefinition.input("b", "B", "Second boolean", DataType.Bool).withDefaultBool(false));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "A OR B", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setBool("result", ctx.getBool("a") || ctx.getBool("b"));
		return ctx.success();
	}
}

export class NotGateNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "not_gate_as";
		def.friendly_name = "NOT Gate (AS)";
		def.description = "Logical NOT of a boolean input";
		def.category = "Control Flow/Logic";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("value", "Value", "Boolean to negate", DataType.Bool).withDefaultBool(false));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "NOT value", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setBool("result", !ctx.getBool("value"));
		return ctx.success();
	}
}

export class GateNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "gate_as";
		def.friendly_name = "Gate (AS)";
		def.description = "Passes execution only when the gate is open";
		def.category = "Control Flow";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("open", "Open", "Gate state", DataType.Bool).withDefaultBool(true));
		def.addPin(PinDefinition.output("passed", "Passed", "Fires when gate is open", DataType.Exec));
		def.addPin(PinDefinition.output("blocked", "Blocked", "Fires when gate is closed", DataType.Exec));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		if (ctx.getBool("open")) {
			ctx.activateExec("passed");
		} else {
			ctx.activateExec("blocked");
		}
		return ctx.finish();
	}
}

const pkg = new NodePackage();
pkg.register(new IfBranchNode());
pkg.register(new CompareNode());
pkg.register(new AndGateNode());
pkg.register(new OrGateNode());
pkg.register(new NotGateNode());
pkg.register(new GateNode());

export function get_nodes(): i64 {
	return pkg.getNodes();
}

export function run(ptr: i32, len: i32): i64 {
	return pkg.run(ptr, len);
}
